src/modules/link.cpp

```diff
@@ -1,22 +1,123 @@
 //
 // Link Module
 //
-
+#include <Arduino.h>
+#include <FastCRC.h>
+#include "../pt/pt.h"
 #include "../modules/link.h"
 #include "../modules/module.h"
+#include "../hal/serial.h"
+
+FastCRC16 CRC16;  // this is the class to create crc16 to match the CCITT crc16 that rpi is using
 
-// Public Variables
-// TODO: Fix this variable name :(
+// Private Variables
 struct link comm;
+static struct pt serialThread;
+static struct pt serialReadThread;
+static struct pt serialSendThread;
+
+uint16_t calc_crc;
+
+// copy the public data packet to this buffer because there might be changes
+// by other modules while serial.cpp is waiting to send this
+data_packet_def update_crc_data_packet;
+
+#ifdef SERIAL_DEBUG
+extern HardwareSerial &serial_debug;
+#endif
+
+static PT_THREAD(serialReadThreadMain(struct pt* pt))
+{
+  PT_BEGIN(pt);
+
+  PT_WAIT_UNTIL(pt, serialHalGetData() != HAL_IN_PROGRESS);
+
+  if (watchdog_exceeded == true)
+  {
+    comm.public_command_packet.alarm_bits |= 1 << ALARM_DROPPED_PACKET;
+    watchdog_exceeded = false;
+    clear_input = true;
+  }
+  else
+  {
+    if ((comm.public_command_packet.sequence_count != sequence_count))
+    {
+#ifdef SERIAL_DEBUG
+      serial_debug.print("unexpected sequence count: ");
+      serial_debug.println(comm.public_command_packet.sequence_count, DEC);
+#endif    
+      comm.public_command_packet.alarm_bits |= 1 << ALARM_CRC_ERROR;
+      clear_input = true;
+    }
+    else
+    {
+      calc_crc = CRC16.ccitt((uint8_t *)&comm.public_command_packet, sizeof(comm.public_command_packet) - 2);
```

Comment:
As mentioned above, can be replaces with (note, pseudo code) 
```
uint16_t crc = 0;
for (int i = 0; i < sizeof(command_packet); i++) {
  crc = _crc_xmodem_update(crc, command_packet[i]); // Note, from the code Ive seen, Noel's python library uses xmodem crc, not ccitt -- just be sure to use the same crc algo and initializer on both sides
}
```
to avoid external dependency

Reply:
I have a function that wraps the avr c function. it is causing issues. maybe the inline assembly straight from the header source..

Reply:
I get many packets through but get many timeouts. I see 50 msec gaps in a couple of places. It might not be noticeable unless specifically running serial debug mode. I will leave fastcrc for the default until this is fixed with standard avr c. here is the function I use which is identical to the one I can test with a standalone application.  checking time it is 2 - 4 msec for either crc in the standalone app. Fastcrc is normally around 1 msec faster in the standalone app. We do not see the 50msec gaps in the standalone app with the avr C crc.

Reply:
I am not sure why, but it is now running without the timeouts. I will check it in with switch set for avr C crc16. I will keep an eye on this in serial debug mode.